README.md
# PubNubFlutter

[![pub package](https://img.shields.io/pub/v/pubnub.svg)](https://pub.dartlang.org/packages/pubnub)

A Flutter plugin to access PubNub functionalities such as publish, subscribe, status and presence..
The plugin allows having more than one client on the plugin side.

## Usage
To use this plugin, add `pubnub_flutter` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

### Example

``` dart
// Instantiate plugin without a filter expression:
 _pubNubFlutter = PubNubFlutter([
      PubNubConfig('client1', 'pub-c-xxxx',
          'sub-cxxxx',
          presenceTimeout: 120),
      PubNubConfig('client2', 'pub-c-xxxx',
          'sub-cxxx'),
    ]);

    _pubNubFlutter.uuid('client1').then((uuid) => print('UUID1: $uuid'));
    _pubNubFlutter.uuid('client2').then((uuid) => print('UUID2: $uuid'));
        
// Instantiate plugin with a filter expression:
_pubNubFlutter = PubNubFlutter([PubNubConfig('client1', 'pub-c-xxxx',
          'sub-cxxxx',
          presenceTimeout: 120,
        uuid: '127c1ab5-fc7f-4c46-8460-3207b6782007',
        filter: 'uuid != "127c1ab5-fc7f-4c46-8460-3207b6782007"']);

// Subscribe to a channel:
_pubNubFlutter.subscribe('client1', ['Olivier-Channel']);

// Unsubscribe from a channel:
_pubNubFlutter.unsubscribe('client1',channel: 'test_channel');

//  Unsubscribe from all channels:
_pubNubFlutter.unsubscribeAll('client1');

// Publish a message to a channel:
_pubNubFlutter.publish('client1',
    {'message': 'Hello World'},
    'test_channel',
);

// Publish a message to a channel passing metadata optional filter expression acts upon:
_pubNubFlutter.publish('client1',
    {'message': 'Hello World'},
    'test_channel',
    metadata: {
        'uuid': '127c1ab5-fc7f-4c46-8460-3207b6782007'
    }
);

// Listen for Messages:
_pubNubFlutter.onMessageReceived.listen((message) => print('Message:$message'));

// Listen for Status:
_pubNubFlutter.onStatusReceived.listen((status) => print('Status:${status.toString()}'));

// Listen to Presence:
_pubNubFlutter.onPresenceReceived.listen((presence) => print('Presence:${presence.toString()}'));

// Listen for Errors:
_pubNubFlutter.onErrorReceived.listen((error) => print('Error:$error'));


```


CHANGELOG.md
## 0.0.1

* Android and iOS PubNub Flutter plugin. Implements subscribe, unsubscribe, publish, filter expressions, presence.


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:pubnub_flutter/pubnub_flutter.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  PubNubFlutter _pubNubFlutter;

  @override
  void initState() {
    super.initState();
    _pubNubFlutter = PubNubFlutter('pub-c-2d1121f9-06c1-4413-8d2e-865f0cfe702a',
        'sub-c-324ae474-ecfd-11e8-91a4-7e00ddddd7aa',
        uuid: '127c1ab5-fc7f-4c46-8460-3207b6782007',
        filter: 'uuid != "127c1ab5-fc7f-4c46-8460-3207b6782007"');

    _pubNubFlutter.uuid().then((uuid) => print('UUID: $uuid'));

    _pubNubFlutter.onStatusReceived
        .listen((status) => print('Status:${status.toString()}'));

    _pubNubFlutter.onPresenceReceived
        .listen((presence) => print('Presence:${presence.toString()}'));

    _pubNubFlutter.onMessageReceived
        .listen((message) => print('Message:$message'));

    _pubNubFlutter.onErrorReceived.listen((error) => print('Error:$error'));
  }

  @override
  void dispose() {
    print('Unsubscribe all');
    _pubNubFlutter.unsubscribeAll();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) => MaterialApp(
        home: Scaffold(
          appBar: AppBar(
            title: const Text('PubNub'),
          ),
          body: Center(
              child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Row(
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  children: <Widget>[
                    FlatButton(
                        color: Colors.black12,
                        onPressed: () {
                          _pubNubFlutter.unsubscribe(
                              channel: 'olivier_channel');
                        },
                        child: Text('Unsubscribe')),
                    FlatButton(
                        color: Colors.black12,
                        onPressed: () {
                          _pubNubFlutter.subscribe(['olivier_channel']);
                        },
                        child: Text('Subscribe')),
                    FlatButton(
                        color: Colors.black12,
                        onPressed: () {
                          _pubNubFlutter.publish(
                            {'message': 'Hello World'},
                            'olivier_channel',
                          );
                          // Below is used to filter the uuid, works in combination with the filter expression in the create method above
                          //metadata: {
                          //  'uuid': '127c1ab5-fc7f-4c46-8460-3207b6782007'
                          //});
                        },
                        child: Text('Send Message'))
                  ])
            ],
          )),
        ),
      );
}


ios/Classes/PubnubFlutterPlugin.h
#import <Flutter/Flutter.h>
#import <PubNub/PubNub.h>

@class MessageStreamHandler;
@class StatusStreamHandler;
@class ErrorStreamHandler;
@class PresenceStreamHandler;

@interface PubnubFlutterPlugin : NSObject<FlutterPlugin>
@property (nonatomic, strong) MessageStreamHandler *messageStreamHandler;
@property (nonatomic, strong) StatusStreamHandler *statusStreamHandler;
@property (nonatomic, strong) PresenceStreamHandler *presenceStreamHandler;
@property (nonatomic, strong) ErrorStreamHandler *errorStreamHandler;
@end

@interface MessageStreamHandler : NSObject<FlutterStreamHandler>
@property (nonatomic, strong) FlutterEventSink eventSink;

- (void) sendMessage:(PNMessageResult *)message;

@end

@interface StatusStreamHandler : NSObject <FlutterStreamHandler>
@property (nonatomic, strong) FlutterEventSink eventSink;

- (void) sendStatus:(PNStatus *)status;

@end

@interface PresenceStreamHandler : NSObject <FlutterStreamHandler>
@property (nonatomic, strong) FlutterEventSink eventSink;

- (void) sendPresence:(PNPresenceEventResult *)presence;

@end

@interface ErrorStreamHandler : NSObject <FlutterStreamHandler>
@property (nonatomic, strong) FlutterEventSink eventSink;

- (void) sendError:(NSDictionary *)error;

@end



ios/Classes/PubnubFlutterPlugin.m
#import "PubnubFlutterPlugin.h"

@interface PubnubFlutterPlugin ()<PNObjectEventListener>

@property (nonatomic, strong) PubNub* client;
@property (nonatomic, strong) PNConfiguration *config;
@end

@implementation PubnubFlutterPlugin

NSString *const PUBNUB_FLUTTER_CHANNEL_NAME = @"flutter.ingenio.com/pubnub_flutter";
NSString *const PUBNUB_MESSAGE_CHANNEL_NAME = @"flutter.ingenio.com/pubnub_message";
NSString *const PUBNUB_STATUS_CHANNEL_NAME = @"flutter.ingenio.com/pubnub_status";
NSString *const PUBNUB_PRESENCE_CHANNEL_NAME = @"flutter.ingenio.com/pubnub_presence";
NSString *const PUBNUB_ERROR_CHANNEL_NAME = @"flutter.ingenio.com/pubnub_error";

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:PUBNUB_FLUTTER_CHANNEL_NAME
                                     binaryMessenger:[registrar messenger]];
    PubnubFlutterPlugin* instance = [[PubnubFlutterPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
    
    // Event channel for streams
    instance.messageStreamHandler = [MessageStreamHandler new];
    
    FlutterEventChannel* messageChannel =
    [FlutterEventChannel eventChannelWithName:PUBNUB_MESSAGE_CHANNEL_NAME
                              binaryMessenger:[registrar messenger]];
    [messageChannel setStreamHandler:instance.messageStreamHandler];
    
    // Event channel for streams
    instance.statusStreamHandler = [StatusStreamHandler new];
    
    FlutterEventChannel* statusChannel =
    [FlutterEventChannel eventChannelWithName:PUBNUB_STATUS_CHANNEL_NAME
                              binaryMessenger:[registrar messenger]];
    [statusChannel setStreamHandler:instance.statusStreamHandler];
    
    // Event channel for streams
    instance.presenceStreamHandler = [PresenceStreamHandler new];
    
    FlutterEventChannel* presenceChannel =
    [FlutterEventChannel eventChannelWithName:PUBNUB_PRESENCE_CHANNEL_NAME
                              binaryMessenger:[registrar messenger]];
    [presenceChannel setStreamHandler:instance.presenceStreamHandler];
    
    instance.errorStreamHandler = [ErrorStreamHandler new];
    
    FlutterEventChannel* errorChannel =
    [FlutterEventChannel eventChannelWithName:PUBNUB_ERROR_CHANNEL_NAME
                              binaryMessenger:[registrar messenger]];
    [errorChannel setStreamHandler:instance.errorStreamHandler];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if  ([@"create" isEqualToString:call.method]) {
        result([self handleCreate:call]);
    } else if  ([@"subscribe" isEqualToString:call.method]) {
        result([self handleSubscribe:call]);
    } else if  ([@"publish" isEqualToString:call.method]) {
        result([self handlePublish:call]);
    } else if  ([@"unsubscribe" isEqualToString:call.method]) {
        result([self handleUnsubscribe:call]);
    } else if  ([@"unsubscribe_all" isEqualToString:call.method]) {
        result([self handleUnsubscribe:call]);
    } else if  ([@"uuid" isEqualToString:call.method]) {     
        result([self handleUUID:call]);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (id) handleUnsubscribe:(FlutterMethodCall*)call {
    NSString *channel = call.arguments[@"channel"];
    
    if(channel) {
        [self.client unsubscribeFromChannels:@[channel] withPresence:NO];
    } else {
        [self.client unsubscribeFromAll];
    }
    
    return NULL;
}

- (id) handleUUID:(FlutterMethodCall*)call {
    
    return self.config.uuid;
}

- (id) handlePublish:(FlutterMethodCall*)call {
    NSString *channel = call.arguments[@"channel"];
    NSDictionary *message = call.arguments[@"message"];
    NSDictionary *metadata = call.arguments[@"metadata"];

    if(channel && message) {
         __weak __typeof(self) weakSelf = self;
        [weakSelf.client publish:message toChannel:channel withMetadata:metadata completion:^(PNPublishStatus *status) {
            __strong __typeof(self) strongSelf = weakSelf;
            [strongSelf handleStatus:status client:strongSelf.client];
        }];
    }
    
    return NULL;
}

- (id) handleCreate:(FlutterMethodCall*)call {
    NSString *publishKey = call.arguments[@"publishKey"];
    NSString *subscribeKey = call.arguments[@"subscribeKey"];
    NSString *uuid = call.arguments[@"uuid"];
    NSString *filter = call.arguments[@"filter"];
    
    if(publishKey && subscribeKey) {
        NSLog(@"Arguments: %@, %@", publishKey, subscribeKey);
       
        self.config =
        [PNConfiguration configurationWithPublishKey:publishKey
                                        subscribeKey:subscribeKey];
        self.config.stripMobilePayload = NO;
        if(uuid) {
            self.config.uuid = uuid;
        }
 
        self.client = [PubNub clientWithConfiguration:self.config];
        
        if(filter) {
            self.client.filterExpression = filter;
        }
        
        [self.client addListener:self];
    }
    
    return NULL;
}

- (id) handleSubscribe:(FlutterMethodCall*)call {

    NSArray *channels = call.arguments[@"channels"];
    
    if(channels) {
        NSLog(@"Arguments: %@", channels);
        
        [self.client subscribeToChannels:channels withPresence:YES];
    }
    
    return NULL;
}

- (void)handleStatus:(PNStatus *)status client:(PubNub*)client {
    if (status.isError) {
        [self.errorStreamHandler sendError:@{@"type":@"state", @"category":  [PubnubFlutterPlugin getCategoryAsNumber:status.category]}];
        NSDictionary *result = @{@"operation":  [PubnubFlutterPlugin getOperationAsNumber:status.operation], @"error": @""};
        [self.errorStreamHandler sendError:result];
        
    } else {
         [self.statusStreamHandler sendStatus:status];
    }
}

#pragma mark - Pubnub delegate methods

- (void)client:(PubNub *)client didReceiveStatus:(PNStatus *)status {
    [self.statusStreamHandler sendStatus:status];
}

- (void)client:(PubNub *)client didReceiveMessage:(PNMessageResult *)message {
    
    // Handle new message stored in message.data.message
    if (![message.data.channel isEqualToString:message.data.subscription]) {
        
        // Message has been received on channel group stored in message.data.subscription.
    }
    else {
        
        // Message has been received on channel stored in message.data.channel.
    }
    
    NSLog(@"Received message: %@ on channel %@ uuid: %@ at %@", message.data.message[@"msg"],
          message.data.channel, message.uuid, message.data.timetoken);
    
    [self.messageStreamHandler sendMessage:message];
}

// New presence event handling.
- (void)client:(PubNub *)client didReceivePresenceEvent:(PNPresenceEventResult *)presence {
    
    if (![presence.data.channel isEqualToString:presence.data.subscription]) {
        
        // Presence event has been received on channel group stored in event.data.subscription.
    }
    else {
        
        // Presence event has been received on channel stored in event.data.channel.
    }
    
    if (![presence.data.presenceEvent isEqualToString:@"state-change"]) {
        
        NSLog(@"%@ \"%@'ed\"\nat: %@ on %@ (Occupancy: %@)", presence.data.presence.uuid,
              presence.data.presenceEvent, presence.data.presence.timetoken, presence.data.channel,
              presence.data.presence.occupancy);
    }
    else {
        
        NSLog(@"%@ changed state at: %@ on %@ to: %@", presence.data.presence.uuid,
              presence.data.presence.timetoken, presence.data.channel, presence.data.presence.state);
    }
    
    [self.presenceStreamHandler sendPresence:presence];
}


+ (NSNumber *) getCategoryAsNumber:(PNStatusCategory) category {
    switch(category) {
            
        case PNUnknownCategory:
            return [NSNumber numberWithInt:0];
        case PNAcknowledgmentCategory:
            return [NSNumber numberWithInt:1];
        case PNAccessDeniedCategory:
            return [NSNumber numberWithInt:2];
        case PNTimeoutCategory:
            return [NSNumber numberWithInt:3];
        case PNNetworkIssuesCategory:
            return [NSNumber numberWithInt:4];
        case PNConnectedCategory:
            return [NSNumber numberWithInt:5];
        case PNReconnectedCategory:
            return [NSNumber numberWithInt:6];
        case PNDisconnectedCategory:
            return [NSNumber numberWithInt:7];
        case PNUnexpectedDisconnectCategory:
            return [NSNumber numberWithInt:8];
        case PNCancelledCategory:
            return [NSNumber numberWithInt:9];
        case PNBadRequestCategory:
            return [NSNumber numberWithInt:10];
        case PNMalformedFilterExpressionCategory:
            return [NSNumber numberWithInt:11];
        case PNMalformedResponseCategory:
            return [NSNumber numberWithInt:12];
        case PNDecryptionErrorCategory:
            return [NSNumber numberWithInt:13];
        case PNTLSConnectionFailedCategory:
            return [NSNumber numberWithInt:14];
        case PNTLSUntrustedCertificateCategory:
            return [NSNumber numberWithInt:15];
        case PNRequestMessageCountExceededCategory:
            return [NSNumber numberWithInt:16];
        case PNRequestURITooLongCategory:
            return [NSNumber numberWithInt:0];
    }
    
    return [NSNumber numberWithInt:0];
}

+ (NSNumber *)  getOperationAsNumber:(PNOperationType) operation {
    switch (operation) {
        
        case PNSubscribeOperation:
            return [NSNumber numberWithInt:1];
        case PNUnsubscribeOperation:
            return [NSNumber numberWithInt:2];
        case PNPublishOperation:
           return [NSNumber numberWithInt:3];
        case PNHistoryOperation:
            return [NSNumber numberWithInt:4];
        case PNHistoryForChannelsOperation:
            return [NSNumber numberWithInt:0];
        case PNDeleteMessageOperation:
            return [NSNumber numberWithInt:6];
        case PNWhereNowOperation:
            return [NSNumber numberWithInt:7];
        case PNHereNowGlobalOperation:
            return [NSNumber numberWithInt:0];
        case PNHereNowForChannelOperation:
            return [NSNumber numberWithInt:0];
        case PNHereNowForChannelGroupOperation:
           return [NSNumber numberWithInt:0];
        case PNHeartbeatOperation:
            return [NSNumber numberWithInt:8];
        case PNSetStateOperation:
            return [NSNumber numberWithInt:9];
        case PNGetStateOperation:
            return [NSNumber numberWithInt:21];
        case PNStateForChannelOperation:
            return [NSNumber numberWithInt:0];
        case PNStateForChannelGroupOperation:
            return [NSNumber numberWithInt:0];
        case PNAddChannelsToGroupOperation:
            return [NSNumber numberWithInt:10];
        case PNRemoveChannelsFromGroupOperation:
            return [NSNumber numberWithInt:11];
        case PNChannelGroupsOperation:
            return [NSNumber numberWithInt:12];
        case PNRemoveGroupOperation:
            return [NSNumber numberWithInt:13];
        case PNChannelsForGroupOperation:
            return [NSNumber numberWithInt:14];
        case PNPushNotificationEnabledChannelsOperation:
            return [NSNumber numberWithInt:15];
        case PNAddPushNotificationsOnChannelsOperation:
            return [NSNumber numberWithInt:16];
        case PNRemovePushNotificationsFromChannelsOperation:
            return [NSNumber numberWithInt:17];;
        case PNRemoveAllPushNotificationsOperation:
            return [NSNumber numberWithInt:18];
        case PNTimeOperation:
            return [NSNumber numberWithInt:19];
    }
    
    return [NSNumber numberWithInt:0];
}
@end


@implementation MessageStreamHandler

- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
    self.eventSink = eventSink;
    return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
    self.eventSink = nil;
    return nil;
}

- (void) sendMessage:(PNMessageResult *)message {
     if(self.eventSink) {
    
         NSDictionary *result = @{@"uuid": message.uuid, @"channel": message.data.channel, @"message": message.data.message};
         self.eventSink(result);
     }
}

@end

@implementation StatusStreamHandler

- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
    self.eventSink = eventSink;
    return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
    self.eventSink = nil;
    return nil;
}

- (void) sendStatus:(PNStatus *)status {
    if(self.eventSink) {
        
        self.eventSink(@{@"category": [PubnubFlutterPlugin getCategoryAsNumber:status.category],@"operation": [PubnubFlutterPlugin getOperationAsNumber:status.operation], @"uuid": status.uuid});
    }
}

@end

@implementation PresenceStreamHandler

- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
    self.eventSink = eventSink;
    return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
    self.eventSink = nil;
    return nil;
}

- (void) sendPresence:(PNPresenceEventResult *)presence {
    if(self.eventSink) {
        
        self.eventSink(@{@"channel": presence.data.channel, @"event": presence.data.presenceEvent, @"uuid": presence.data.presence.uuid, @"occupancy": presence.data.presence.occupancy});
    }
}

@end

@implementation ErrorStreamHandler

- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
    self.eventSink = eventSink;
    return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
    self.eventSink = nil;
    return nil;
}

- (void) sendError:(NSDictionary *)error {
    if(self.eventSink) {
        self.eventSink(error);
    }
}

@end



lib/pubnub_flutter.dart
import 'dart:async';

import 'package:flutter/services.dart';

/// PubNub Plugin. This plugin is not intended to implement all PubNub functionalities but rather take a minimal approach
/// for solving most general use cases
/// Main areas covered by the plugin are
/// - Instantiate the plugin passing the required PubNub authentication information
/// - Pass a filter expression when instantiating the plugin
/// - Subscribe to one or more channels
/// - Unsubscribe from one channel or all channels
/// - Publish a message to a channel
/// - Retrieve UUID if was not set during the plugin instantiation
/// {@tool sample}
///
/// Instantiate plugin without a filter expression:
///
/// ```dart
/// _pubNubFlutter = PubNubFlutter('pub-c-2d1121f9-06c1-4413-8d2e-0000000000',
///        'sub-c-324ae474-ecfd-11e8-91a4-00000000000',
///        uuid: '127c1ab5-fc7f-4c46-8460-3207b6782007');
/// ```
/// Instantiate plugin with a filter expression:
///
/// ```dart
/// _pubNubFlutter = PubNubFlutter('pub-c-2d1121f9-06c1-4413-8d2e-0000000000',
///        'sub-c-324ae474-ecfd-11e8-91a4-00000000000',
///        uuid: '127c1ab5-fc7f-4c46-8460-3207b6782007',
///        filter: 'uuid != "127c1ab5-fc7f-4c46-8460-3207b6782007"');
/// ```
///
/// Subscribe to a channel:
///
/// ``` dart
/// _pubNubFlutter.subscribe(['test_channel']);
/// ```
///
/// Unsubscribe from a channel:
///
/// ``` dart
/// _pubNubFlutter.unsubscribe(channel: 'test_channel');
/// ```
///
///  Unsubscribe from all channels:
///
/// ``` dart
/// _pubNubFlutter.unsubscribeAll();
/// ```
///
/// Publish a message to a channel:
///
/// ``` dart
///    _pubNubFlutter.publish(
///                            {'message': 'Hello World'},
///                            'test_channel',
///                          );
/// ```
///
/// Publish a message to a channel passing metadata optional filter expression acts upon:
///
/// ``` dart
///    _pubNubFlutter.publish(
///                            {'message': 'Hello World'},
///                            'test_channel',
///                            metadata: {
///                             'uuid': '127c1ab5-fc7f-4c46-8460-3207b6782007'
///                           }
///                          );
/// ```
///
/// Listen for Messages:
///
/// ``` dart
/// _pubNubFlutter.onMessageReceived
///        .listen((message) => print('Message:$message'));
/// ```
///
/// Listen for Status:
///
/// ``` dart
///  _pubNubFlutter.onStatusReceived
///        .listen((status) => print('Status:${status.toString()}'));
/// ```
/// Listen to Presence:
///
/// ``` dart
/// _pubNubFlutter.onPresenceReceived
///        .listen((presence) => print('Presence:${presence.toString()}'));
/// ```
///
/// Listen for Errors:
///
/// ``` dart
/// _pubNubFlutter.onErrorReceived.listen((error) => print('Error:$error'));
/// ```
///
///  {@end-tool}
class PubNubFlutter {
  MethodChannel _channel;
  EventChannel _messageChannel;
  EventChannel _statusChannel;
  EventChannel _presenceChannel;
  EventChannel _errorChannel;

  Stream<Map> _onMessageReceived;
  Stream<Map> _onStatusReceived;
  Stream<Map> _onPresenceReceived;
  Stream<Map> _onErrorReceived;

  /// Create the plugin, UUID and filter expressions are optional and can be used for tracking purposes and filtering purposes, for instance can disable getting messages on the same UUID.
  PubNubFlutter(String publishKey, String subscribeKey,
      {String uuid, String filter}) {
    _channel = MethodChannel('flutter.ingenio.com/pubnub_flutter');
    _messageChannel = const EventChannel('flutter.ingenio.com/pubnub_message');
    _statusChannel = const EventChannel('flutter.ingenio.com/pubnub_status');
    _presenceChannel =
        const EventChannel('flutter.ingenio.com/pubnub_presence');
    _errorChannel = const EventChannel('flutter.ingenio.com/pubnub_error');

    var args = {'publishKey': publishKey, 'subscribeKey': subscribeKey};
    if (uuid != null) {
      args['uuid'] = uuid;
    }
    if (filter != null) {
      args['filter'] = filter;
    }
    _channel.invokeMethod('create', args);
  }

  /// Subscribe to a list of channels
  Future<void> subscribe(List<String> channels) async {
    await _channel.invokeMethod('subscribe', {'channels': channels});
    return;
  }

  /// Publishes a message on a specified channel, some metadata can be passed and used in conjunction with filter expressions
  Future<void> publish(Map message, String channel, {Map metadata}) async {
    Map args = {'message': message, 'channel': channel};

    if (metadata != null) {
      args['metadata'] = metadata;
    }

    return await _channel.invokeMethod('publish', args);
  }

  /// Unsubscribes from a single channel
  Future<void> unsubscribe({String channel}) async {
    return await _channel.invokeMethod('unsubscribe', {'channel': channel});
  }

  /// Unsubscribes from all channels
  Future<void> unsubscribeAll() async {
    return await _channel.invokeMethod('unsubscribe');
  }

  /// Get the UUID configured for PubNub. Note that when the UUID is passed  in the plugin creation, the returned UUID is the same
  /// If the UUID has not been passed in the plugin creation, then PubNub assigns a new UUID. This may be important for tracking how many devices/clients are using the API and
  /// may impact how much the service costs
  Future<String> uuid() async {
    return await _channel.invokeMethod('uuid');
  }

  /// Fires whenever the a message is received.
  Stream<Map> get onMessageReceived {
    if (_onMessageReceived == null) {
      _onMessageReceived = _messageChannel
          .receiveBroadcastStream()
          .map((dynamic event) => _parseMessage(event));
    }
    return _onMessageReceived;
  }

  /// Fires whenever the status changes.
  Stream<Map> get onStatusReceived {
    if (_onStatusReceived == null) {
      _onStatusReceived = _statusChannel
          .receiveBroadcastStream()
          .map((dynamic event) => _parseStatus(event));
    }
    return _onStatusReceived;
  }

  /// Fires whenever the presence changes.
  Stream<Map> get onPresenceReceived {
    if (_onPresenceReceived == null) {
      _onPresenceReceived = _presenceChannel
          .receiveBroadcastStream()
          .map((dynamic event) => _parsePresence(event));
    }
    return _onPresenceReceived;
  }

  /// Fires whenever an error is received.
  Stream<Map> get onErrorReceived {
    if (_onErrorReceived == null) {
      _onErrorReceived = _errorChannel
          .receiveBroadcastStream()
          .map((dynamic event) => _parseError(event));
    }
    return _onErrorReceived;
  }

  /// Fires whenever a status is received.
  Map _parseStatus(Map status) {
    int category = status['category'];
    status['category'] = PNStatusCategory.values[category];
    int operation = status['operation'];
    status['operation'] = PNOperationType.values[operation];
    return status;
  }

  /// Fires whenever presence is received
  Map _parsePresence(Map presence) {
    return presence;
  }

  /// Fires whenever a PubNub error is received
  Map _parseError(Map error) {
    int operation = error['operation'];
    error['operation'] = PNOperationType.values[operation];
    return error;
  }

  /// Fires whenever a message is received
  Map _parseMessage(Map message) {
    return message;
  }
}

/// Values for the status category. Not this is an intersection of both iOS and Android enums as both have different values
enum PNStatusCategory {
  PNUnknownCategory,
  PNAcknowledgmentCategory,
  PNAccessDeniedCategory,
  PNTimeoutCategory,
  PNNetworkIssuesCategory,
  PNConnectedCategory,
  PNReconnectedCategory,
  PNDisconnectedCategory,
  PNUnexpectedDisconnectCategory,
  PNCancelledCategory,
  PNBadRequestCategory,
  PNMalformedFilterExpressionCategory,
  PNMalformedResponseCategory,
  PNDecryptionErrorCategory,
  PNTLSConnectionFailedCategory,
  PNTLSUntrustedCertificateCategory,
  PNRequestMessageCountExceededCategory,
}

/// Operation type coming back in the status
enum PNOperationType {
  PNUnknownOperation,
  PNSubscribeOperation,
  PNUnsubscribeOperation,
  PNPublishOperation,
  PNHistoryOperation,
  PNFetchMessagesOperation,
  PNDeleteMessagesOperation,
  PNWhereNowOperation,
  PNHeartbeatOperation,
  PNSetStateOperation,
  PNAddChannelsToGroupOperation,
  PNRemoveChannelsFromGroupOperation,
  PNChannelGroupsOperation,
  PNRemoveGroupOperation,
  PNChannelsForGroupOperation,
  PNPushNotificationEnabledChannelsOperation,
  PNAddPushNotificationsOnChannelsOperation,
  PNRemovePushNotificationsFromChannelsOperation,
  PNRemoveAllPushNotificationsOperation,
  PNTimeOperation,
}
